File: face/geomajas-face-dojo/client/src/main/resources/geomajas/controller/ZoomOnClickController.js
/*
 * This is part of Geomajas, a GIS framework, http://www.geomajas.org/.
 *
 * Copyright 2008-2011 Geosparc nv, http://www.geosparc.com/, Belgium.
 *
 * The program is available in open source according to the GNU Affero
 * General Public License. All contributions in this program are covered
 * by the Geomajas Contributors License Agreement. For full licensing
 * details, see LICENSE.txt in the project root.
 */

dojo.provide("geomajas.controller.ZoomOnClickController");
dojo.require("geomajas.event.MouseListener");
dojo.require("geomajas.spatial.transform.WorldViewTransformation");

dojo.declare("ZoomOnClickController", MouseListener, {

	/**
	 * @fileoverview MouseListener for zooming by clicking.
	 * @class MouseListener implementation for zooming on a click event.
	 * Whether it zooms in or out, depends on the given scaleFactor.
	 * @author Pieter De Graef
	 *
	 * @constructor
	 * @extends MouseListener
	 * @param mapView Reference to a MapWidget's viewing object.
	 * @param scaleFactor Determines the difference in scale.
	 */
	constructor : function (mapView, scaleFactor, name) {
		/** Reference to a MapWidget's MapView object. */
		this.mapView = mapView;
		
		/** The scale-delta. Bigger then 1 zooms in, smaller then 1 zooms out. */
		this.scaleFactor = scaleFactor;
		
		this.name = name;
	},

	/**
	 * Return a unique name for this MouseListener.
	 */
	getName : function () {
		if (this.name) {
			return this.name;
		}
		return "ZoomOnClickController";
	},

	/**
	 * Zooms in or out around the position where the user clicked.
	 * @param event A HtmlMouseEvent object.
	 */
	mouseClicked : function (/*HtmlMouseEvent*/event) {
		var transform = new WorldViewTransformation(this.mapView);

		var viewPosition = event.getPosition();
		var worldPosition = transform.viewPointToWorld(viewPosition);
		
		this.mapView.getCamera().setPosition (worldPosition.getX(), worldPosition.getY());
		this.mapView.scale(this.scaleFactor, geomajas.ZoomOption.ZOOM_OPTION_LEVEL_CHANGE);
	}

});

File: client/geomajas-client-smartgwt/client/src/main/java/org/geomajas/smartgwt/client/controller/ZoomOnClickController.java
/*
 * This is part of Geomajas, a GIS framework, http://www.geomajas.org/.
 *
 * Copyright 2008-2013 Geosparc nv, http://www.geosparc.com/, Belgium.
 *
 * The program is available in open source according to the GNU Affero
 * General Public License. All contributions in this program are covered
 * by the Geomajas Contributors License Agreement. For full licensing
 * details, see LICENSE.txt in the project root.
 */
package org.geomajas.smartgwt.client.controller;

import org.geomajas.geometry.Coordinate;
import org.geomajas.smartgwt.client.map.MapView;
import org.geomajas.smartgwt.client.spatial.WorldViewTransformer;
import org.geomajas.smartgwt.client.widget.MapWidget;

import com.google.gwt.event.dom.client.MouseUpEvent;

/**
 * Controller for zooming when clicking.
 *
 * @author Joachim Van der Auwera
 */
public class ZoomOnClickController extends AbstractGraphicsController {

	private double zoomFactor;

	public ZoomOnClickController(MapWidget mapWidget, double zoomFactor) {
		super(mapWidget);
		this.zoomFactor = zoomFactor;
	}

	@Override
	public void onMouseUp(MouseUpEvent event) {
		MapView mapView = mapWidget.getMapModel().getMapView();
		WorldViewTransformer transformer = mapView.getWorldViewTransformer();
		Coordinate viewPosition = getScreenPosition(event);
		Coordinate worldPosition = transformer.viewToWorld(viewPosition);
		mapView.setCenterPosition(worldPosition);
		mapView.scale(zoomFactor, MapView.ZoomOption.LEVEL_CHANGE);
	}

	/**
	 * Set zoom factor. Should be a positive number, &gt;1 for zooming in, &lt;1 for zooming out.
	 *
	 * @param zoomFactor zoom factor
	 */
	public void setZoomFactor(double zoomFactor) {
		this.zoomFactor = zoomFactor;
	}
}
